core/src/main/java/fj/Equal.java

```diff
@@ -424,16 +334,8 @@ public Boolean f(final P3<A, B, C> p2) {
    */
   public static <A, B, C, D> Equal<P4<A, B, C, D>> p4Equal(final Equal<A> ea, final Equal<B> eb, final Equal<C> ec,
                                                            final Equal<D> ed) {
-    return new Equal<P4<A, B, C, D>>(new F<P4<A, B, C, D>, F<P4<A, B, C, D>, Boolean>>() {
-      public F<P4<A, B, C, D>, Boolean> f(final P4<A, B, C, D> p1) {
-        return new F<P4<A, B, C, D>, Boolean>() {
-          public Boolean f(final P4<A, B, C, D> p2) {
-            return ea.eq(p1._1(), p2._1()) && eb.eq(p1._2(), p2._2()) && ec.eq(p1._3(), p2._3()) &&
-                   ed.eq(p1._4(), p2._4());
-          }
-        };
-      }
-    });
+    return new Equal<P4<A, B, C, D>>(p1 -> p2 -> ea.eq(p1._1(), p2._1()) && eb.eq(p1._2(), p2._2()) && ec.eq(p1._3(), p2._3()) &&
+           ed.eq(p1._4(), p2._4()));
```

Comment:
The use of Equal.equal would really help here to take advantage of type inference.